throughput/spam.go

```diff
@@ -0,0 +1,454 @@
+	trials := s.txsPerSecond * 60 * 2 // sender/receiver
+	unique := set.NewSet[uint64](trials)
+	for i := 0; i < trials; i++ {
+		unique.Add(zz.Uint64())
+	}
+	utils.Outf("{{blue}}unique participants expected every 60s:{{/}} %d\n", unique.Len())
+}
+
+func (s *Spammer) logStats(cctx context.Context, txIssuer *issuer) *time.Ticker {
+	// Log stats
+	t := time.NewTicker(1 * time.Second) // ensure no duplicates created
+	var psent int64
+	go func() {
+		for {
+			select {
+			case <-t.C:
+				current := sent.Load()
+				l.Lock()
+				if totalTxs > 0 {
+					unitPrices, err := txIssuer.cli.UnitPrices(cctx, false)
+					if err != nil {
+						continue
+					}
+					utils.Outf(
+						"{{yellow}}txs seen:{{/}} %d {{yellow}}success rate:{{/}} %.2f%% {{yellow}}inflight:{{/}} %d {{yellow}}issued/s:{{/}} %d {{yellow}}unit prices:{{/}} [%s]\n", //nolint:lll
+						totalTxs,
+						float64(confirmedTxs)/float64(totalTxs)*100,
+						inflight.Load(),
+						current-psent,
+						unitPrices,
+					)
+				}
+				l.Unlock()
+				psent = current
+			case <-cctx.Done():
+				return
+			}
+		}
+	}()
+	// ensure to stop the ticker when done
+	return t
+}
+
+// createIssuer creates an [numClients] transaction issuers for each URI in [uris]
+func (s *Spammer) createIssuers(parser chain.Parser) ([]*issuer, error) {
+	issuers := []*issuer{}
+	for i := 0; i < len(s.uris); i++ {
+		for j := 0; j < s.numClients; j++ {
+			cli := jsonrpc.NewJSONRPCClient(s.uris[i])
+			webSocketClient, err := ws.NewWebSocketClient(s.uris[i], ws.DefaultHandshakeTimeout, pubsub.MaxPendingMessages, pubsub.MaxReadMessageSize) // we write the max read
+			if err != nil {
+				return nil, err
+			}
+			issuer := &issuer{i: len(issuers), cli: cli, ws: webSocketClient, parser: parser, uri: s.uris[i]}
+			issuers = append(issuers, issuer)
+		}
+	}
+	return issuers, nil
+}
+
+func (s *Spammer) distributeFunds(ctx context.Context, cli *jsonrpc.JSONRPCClient, parser chain.Parser, feePerTx uint64, sh SpamHelper) ([]*auth.PrivateKey, map[codec.Address]uint64, []chain.AuthFactory, error) {
+	withholding := feePerTx * uint64(s.numAccounts)
+	if s.balance < withholding {
+		return nil, nil, nil, fmt.Errorf("insufficient funds (have=%d need=%d)", s.balance, withholding)
+	}
+
+	distAmount := (s.balance - withholding) / uint64(s.numAccounts)
+
+	utils.Outf("{{yellow}}distributing funds to each account{{/}}\n")
+
+	funds := map[codec.Address]uint64{}
+	accounts := make([]*auth.PrivateKey, s.numAccounts)
+	factories := make([]chain.AuthFactory, s.numAccounts)
+
+	factory, err := auth.GetFactory(s.key)
+	if err != nil {
+		return nil, nil, nil, err
+	}
+
+	webSocketClient, err := ws.NewWebSocketClient(s.uris[0], ws.DefaultHandshakeTimeout, pubsub.MaxPendingMessages, pubsub.MaxReadMessageSize) // we write the max read
+	if err != nil {
+		return nil, nil, nil, err
+	}
+	p := &pacer{ws: webSocketClient}
+	go p.Run(ctx, s.minTxsPerSecond)
+	// This helps from hanging
+	time.Sleep(3 * time.Second)
```

Comment:
need to debug, but something about this logic was finicky with the previous code

Reply:
Could you add a TODO with enough detail for the auto-generated GitHub issue so that we can follow up?